### PR TITLE
#362: Brought up code coverage for Java APIs and more

### PR DESCRIPTION
**Pull request by akara**
Thanks for your pull request.  Please review the following guidelines.

- [X] Title includes issue id.
- [X] Description of the change added.
- [X] Commits are squashed.
- [X] Tests added.
- [ ] Documentation added/updated.
- [X] Also please review [CONTRIBUTING.md](https://github.com/paypal/squbs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: squbs-ext/src/test/java/org/squbs/endpoint/DummyLocalhost2Resolver.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ */
+package org.squbs.endpoint;
+
+import org.squbs.env.DEV;
+import org.squbs.env.Default;
+import org.squbs.env.Environment;
+
+import java.util.Optional;
+
+public class DummyLocalhost2Resolver extends AbstractEndpointResolver {
```
Comment (anilgursel) [inline squbs-ext/src/test/java/org/squbs/endpoint/DummyLocalhost2Resolver.java]: Can we rename this as `DummyLocalhostJavaResolver` instead?  Or move it under `org.squbs.endpoint.japi` and just name `DummyLocalhostResolver`?

File: squbs-ext/src/test/java/org/squbs/endpoint/EndpointTest.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+public class EndpointTest {
+
+    private static final ActorSystem system = ActorSystem.create("EndpointTest");
+
+    @AfterClass
+    public static void afterAll() {
+        system.terminate();
+    }
+
+    private void clear() {
```
Comment (anilgursel) [inline squbs-ext/src/test/java/org/squbs/endpoint/EndpointTest.java]: Instead of calling this in each test mode, we should annotate it with `@After` to run after each test.

File: squbs-ext/src/test/java/org/squbs/endpoint/EndpointTest.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    public void registerLambdaResolver() {
+        EndpointResolverRegistry.get(system).register("RogueResolver",
+                (svcName, env) -> Optional.of(Endpoint.create("http://myrogueservice.com")));
+        assertEquals(1, EndpointResolverRegistry.get(system).endpointResolvers().size());
+        assertEquals("RogueResolver", EndpointResolverRegistry.get(system).endpointResolvers().head().name());
+        clear();
+    }
+
+    @Test
+    public void registerResolveSSLResolver() throws NoSuchAlgorithmException {
```
Comment (anilgursel) [inline squbs-ext/src/test/java/org/squbs/endpoint/EndpointTest.java]: Since you already catch `NoSuchAlgorithmException`, do we still need the `throws NoSuchAlgorithmException`?

File: squbs-ext/src/test/java/org/squbs/endpoint/EndpointTest.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        EndpointResolverRegistry.get(system).register("RogueResolver",
+                (svcName, env) -> Optional.of(Endpoint.create("http://myrogueservice.com")));
+        Option<EndpointResolver> resolver =
+                EndpointResolverRegistry.get(system).route("abcService", Default.value());
+        assertEquals("RogueResolver", resolver.get().name());
+        assertEquals("http://myrogueservice.com",
+                resolver.get().resolve("abcService", Default.value()).get().uri().toString());
+        clear();
+    }
+}
```
Comment (anilgursel) [inline squbs-ext/src/test/java/org/squbs/endpoint/EndpointTest.java]: We should probably also have a tests that checks the ordering when two resolvers are registered.  Also, one that checks the ordering when one is registered with lambda API and the other with implementing the abstract class.

File: squbs-ext/src/test/java/org/squbs/env/EnvironmentTest.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    public static void afterAll() {
+        system.terminate();
+    }
+
+    @Test
+    public void resolveDefault() {
+        EnvironmentResolverRegistry.get(system).register(DummyProdEnvironmentResolver.get());
+        Environment env = EnvironmentResolverRegistry.get(system).resolve();
+        assertEquals(PROD.value(), env);
+        EnvTestHelper.clearRegistries(system);
```
Comment (anilgursel) [inline squbs-ext/src/test/java/org/squbs/env/EnvironmentTest.java]: Instead of calling this in each test, we should probably do it with `@After`?

File: squbs-ext/src/test/scala/org/squbs/endpoint/EndpointSpec.scala [hunk 6]
```diff
@@ -63,13 +82,26 @@ class EndpointSpec extends TestKit(ActorSystem("EndpointSpec")) with FlatSpecLik
     resolver.value.resolve("abcService") should be (Some(Endpoint("http://localhost:8080")))
   }
 
+  it should "resolve the endpoint with resolver lambda" in {
+    EndpointResolverRegistry(system).register("RogueResolver", (_, _) => Some(Endpoint("http://myrogueservice.com")))
+    val resolver = EndpointResolverRegistry(system).route("abcService")
```
Comment (anilgursel) [inline squbs-ext/src/test/scala/org/squbs/endpoint/EndpointSpec.scala]: Just thinking out loud.  Should we rename `route` as `findResolver`?  Somehow `route` does not sound right to me. 

Moreover, AFAIR, it is used only in tests.  Should we consider making the method package private?  I feel like we adding it to the public API without an actual use case.

File: squbs-ext/src/test/scala/org/squbs/env/EnvironmentSpec.scala [hunk 7]
```diff
@@ -57,4 +57,16 @@ with Matchers with BeforeAndAfterEach{
     EnvironmentResolverRegistry(system).resolve should be (PROD)
   }
 
+  it should "just warn and not throw exceptions when unregister non-existing resolver" in {
```
Comment (anilgursel) [inline squbs-ext/src/test/scala/org/squbs/env/EnvironmentSpec.scala]: `it should "just warn and not throw exceptions when non-existing resolver is unregistered"` ?

Actually, since we do not assert "just warn" part, I would reword it as

`it should "just warn and not throw exceptions when unregister non-existing resolver"`

File: squbs-ext/src/test/scala/org/squbs/endpoint/EndpointSpec.scala (hunk 6 above)
Comment (akara) [inline squbs-ext/src/test/scala/org/squbs/endpoint/EndpointSpec.scala]: Will do. It did not sound right to me, either.

File: squbs-ext/src/test/java/org/squbs/endpoint/EndpointTest.java (hunk 4 above)
Comment (akara) [inline squbs-ext/src/test/java/org/squbs/endpoint/EndpointTest.java]: Will do. I saw you have those in the Scala tests already so I was just intending to test the Java API. But sure, will do that.

File: squbs-ext/src/test/scala/org/squbs/env/EnvironmentSpec.scala (hunk 7 above)
Comment (akara) [inline squbs-ext/src/test/scala/org/squbs/env/EnvironmentSpec.scala]: Did you mean `it should "not throw exceptions when unregister non-existing resolver"`?

Otherwise the suggested text is the same as what's in there.
